# Imbalance mapping analysis

Let's get down to it! This module is relatively easy to use, but the explanation are slightly more complex.

## Practice data

As always, `sihnpy` comes with data from the PREVENT-AD to be used as practice. In this case, **structural MRI data, preprocessed with FreeSurfer** is provided (see the {ref}`PREVENT-AD preprocessing documentation for more info <0.pad_data/datasets_usage:Additional information on brain imaging preprocessing>`).

## Imbalance mapping

### 1. Preparing the data

**Imbalance mapping** only requires:
* A `pandas.DataFrame` where 
    * The index is the participants' IDs
    * The columns contain the variables you want to run the **imbalance mapping** on (e.g., thickness/volume).

If you already have your data, you can skip ahead to the {ref}`next section <4.imbalance_mapping/imb_module:2. Imbalance mapping>`.

Loading the data is super easy:


In [1]:
from sihnpy.datasets import pad_imb_input

volume_data, thickness_data, aseg_data = pad_imb_input()
thickness_data

session      run  ctx_lh_bankssts_thickness  \
participant_id                                                 
sub-1000173     ses-FU12  run-001                      2.153   
sub-1002928     ses-BL00  run-001                      2.462   
sub-1004359     ses-BL00  run-001                      2.246   
sub-1004359     ses-FU12  run-001                      2.238   
sub-1016072     ses-BL00  run-001                      2.213   
...                  ...      ...                        ...   
sub-9930257     ses-BL00  run-001                      2.435   
sub-9931234     ses-BL00  run-001                      2.461   
sub-9931234     ses-FU12  run-001                      2.403   
sub-9939055     ses-BL00  run-001                      2.332   
sub-9939055     ses-FU12  run-001                      2.312   

                ctx_lh_caudalanteriorcingulate_thickness  \
participant_id                                             
sub-1000173                                        2.334   
sub-1002928                                        2.403   
sub-1004359                                        2.276   
sub-1004359                                        2.339   
sub-1016072                                        2.544   
...                                                  ...   
sub-9930257                                        2.476   
sub-9931234                                        2.166   
sub-9931234                                        2.194   
sub-9939055                                        2.743   
sub-9939055                                        2.611   

                ctx_lh_caudalmiddlefrontal_thickness  ctx_lh_cuneus_thickness  \
participant_id                                                                  
sub-1000173                                    2.451                    1.767   
sub-1002928                                    2.449                    1.735   
sub-1004359                                    2.346                    1.815   
sub-1004359                                    2.349                    1.827   
sub-1016072                                    2.327                    1.662   
...                                              ...                      ...   
sub-9930257                                    2.511                    1.893   
sub-9931234                                    2.526                    1.755   
sub-9931234                                    2.536                    1.756   
sub-9939055                                    2.460                    1.648   
sub-9939055                                    2.456                    1.622   

                ctx_lh_entorhinal_thickness  ctx_lh_fusiform_thickness  \
participant_id                                                           
sub-1000173                           3.202                      2.452   
sub-1002928                           3.316                      2.664   
sub-1004359                           3.332                      2.728   
sub-1004359                           3.449                      2.674   
sub-1016072                           3.310                      2.580   
...                                     ...                        ...   
sub-9930257                           2.903                      2.557   
sub-9931234                           3.143                      2.618   
sub-9931234                           3.216                      2.594   
sub-9939055                           3.010                      2.609   
sub-9939055                           2.960                      2.583   

                ctx_lh_inferiorparietal_thickness  \
participant_id                                      
sub-1000173                                 2.161   
sub-1002928                                 2.253   
sub-1004359                                 2.388   
sub-1004359                                 2.403   
sub-1016072                                 2.295   
...            

As I mentioned {ref}`in the introduction <4.imbalance_mapping/imb_intro:Data type>`, `sihnpy` assumes that the data has a single row for each participant. However, to give you more practice opportunities, `sihnpy` data for this module comes with two visits: BL00 and FU12. For now, we will restrict the data to only the baseline visits and drop the "session" and "run" variables as they aren't needed.

In [2]:
thickness_data_final = thickness_data[thickness_data['session'] == 'ses-BL00'].drop(labels=['session', 'run'], axis=1)
thickness_data_final #We get the 68 columns of data needed

ctx_lh_bankssts_thickness  \
participant_id                              
sub-1002928                         2.462   
sub-1004359                         2.246   
sub-1016072                         2.213   
sub-1031654                         2.455   
sub-1072774                         2.403   
...                                   ...   
sub-9889544                         2.438   
sub-9909448                         2.502   
sub-9930257                         2.435   
sub-9931234                         2.461   
sub-9939055                         2.332   

                ctx_lh_caudalanteriorcingulate_thickness  \
participant_id                                             
sub-1002928                                        2.403   
sub-1004359                                        2.276   
sub-1016072                                        2.544   
sub-1031654                                        2.340   
sub-1072774                                        2.480   
...                                                  ...   
sub-9889544                                        2.438   
sub-9909448                                        2.062   
sub-9930257                                        2.476   
sub-9931234                                        2.166   
sub-9939055                                        2.743   

                ctx_lh_caudalmiddlefrontal_thickness  ctx_lh_cuneus_thickness  \
participant_id                                                                  
sub-1002928                                    2.449                    1.735   
sub-1004359                                    2.346                    1.815   
sub-1016072                                    2.327                    1.662   
sub-1031654                                    2.355                    1.842   
sub-1072774                                    2.485                    1.752   
...                                              ...                      ...   
sub-9889544                                    2.441                    1.838   
sub-9909448                                    2.359                    1.750   
sub-9930257                                    2.511                    1.893   
sub-9931234                                    2.526                    1.755   
sub-9939055                                    2.460                    1.648   

                ctx_lh_entorhinal_thickness  ctx_lh_fusiform_thickness  \
participant_id                                                           
sub-1002928                           3.316                      2.664   
sub-1004359                           3.332                      2.728   
sub-1016072                           3.310                      2.580   
sub-1031654                           3.381                      2.833   
sub-1072774                           3.443                      2.672   
...                                     ...                        ...   
sub-9889544                           3.306                      2.613   
sub-9909448                           3.127                      2.642   
sub-9930257                           2.903                      2.557   
sub-9931234                           3.143                      2.618   
sub-9939055                           3.010                      2.609   

                ctx_lh_inferiorparietal_thickness  \
participant_id                                      
sub-1002928                                 2.253   
sub-1004359                                 2.388   
sub-1016072                                 2.295   
sub-1031654                                 2.290   
sub-1072774                                 2.356   
...                                           ...   
sub-9889544                                 2.315   
sub-9909448                                 2.314   
sub-9930257                                 2.397   
sub-9931234                                 2.383

### 2. Imbalance mapping

Ok now is the fun part: the **imbalance mapping**. The code behind is complex, but the application is simple. The function only requires two input:
* A `pandas.DataFrame` with the index as the participant IDs and the columns being the variables we want to map.
* The `type` of orthogonal distances we want to get.

The `type` of distances is simply whether we want **absolute** (`abs`) distances or **signed** (`sign`) distances from the regression slope. In the **absolute** version, we don't care whether the fitted value is above or below the diagonal: we just want to know how far the value is. In the **signed** version, we want to know whether the fitted value was overestimated (i.e., above the regression line) or underestimated (i.e., below the regression line). Whether you want **absolute** or **signed** distances is really up to you. In the original article by Nadig et al. (2021)[^Nadig_2021], the **absolute** residuals were used. Here is the code to run it:

In [3]:
from sihnpy import imbalance_mapping as imb

residual_array = imb.imbalance_mapping(data=thickness_data_final, type='abs')

Computing region 1/68
Computing region 2/68
Computing region 3/68
Computing region 4/68
Computing region 5/68
Computing region 6/68
Computing region 7/68
Computing region 8/68
Computing region 9/68
Computing region 10/68
Computing region 11/68
Computing region 12/68
Computing region 13/68
Computing region 14/68
Computing region 15/68
Computing region 16/68
Computing region 17/68
Computing region 18/68
Computing region 19/68
Computing region 20/68
Computing region 21/68
Computing region 22/68
Computing region 23/68
Computing region 24/68
Computing region 25/68
Computing region 26/68
Computing region 27/68
Computing region 28/68
Computing region 29/68
Computing region 30/68
Computing region 31/68
Computing region 32/68
Computing region 33/68
Computing region 34/68
Computing region 35/68
Computing region 36/68
Computing region 37/68
Computing region 38/68
Computing region 39/68
Computing region 40/68
Computing region 41/68
Computing region 42/68
Computing region 43/68
Computing region 44/

So, what did we do, and what did we get? 

In this step, we do the equivalent of step B) and C) from {ref}`Nadig et al. (2021)'s figure <4.imbalance_mapping/imb_intro:Definitions>`: we compute the structural covariance and extract the **orthogonal distances** for each participant. This is store in a massive 3D matrix (well... I guess it's a rectangle at this point):

In [4]:
residual_array.shape

(68, 68, 306)

Here's a simplified illustration of what it would look like if we illustrated this:

```{figure} ../images/imb/resid_array_rep.png
:name: Residual array - Illustration
:scale: 30
:align: center
```
<br>

In this cube, each square matrix represent the orthogonal distances from the regression of the values of two regions for a single participant. The diagonal is removed (in the image it is whited out) as the correlation of the same column twice would give a perfect correlation.

While the shape of the image is much smaller than the actual matrix (68 regions X 68 regions X 306 participants), the principle is exactly the same.

Once this 3D matrix is calculated, we are already ready to compute our measures.

### 3. Imbalance statistics

In the original code by Nadig et al. (2021)[^Nadig_2021] proposes two measures: computing an average imbalance by person (i.e., average imbalance across all edges) and an average imbalance by region at the group-level (i.e., how much imbalance does the entorhinal cortex show for example). `sihnpy` integrates both measures, but also outputs an average imbalance by person by region (i.e., how much imbalance on average the entorhinal cortex shows in person Y). Below is a quick illustration of these measures:

```{figure} ../images/imb/imb_measures.png
:name: Imbalance mapping measures
:scale: 30
:align: center
```

These measures in `sihnpy` can be calculated very easily with a single function taking two arguments:
* The original `pandas.DataFrame` used to compute the **imbalance mapping** (we use it to grab the participant IDs and the region names)
* The cube of orthogonal distances we computed in the previous step

In [5]:
avg_imb_by_region, avg_imb_by_person, avg_imb_by_pers_by_region = imb.imbalance_stats(data=thickness_data_final, residual_array=residual_array)
avg_imb_by_pers_by_region

ctx_lh_bankssts_thickness  \
participant_id                              
sub-1002928                      0.050868   
sub-1004359                      0.108628   
sub-1016072                      0.153541   
sub-1031654                      0.059193   
sub-1072774                      0.042627   
...                                   ...   
sub-9889544                      0.047168   
sub-9909448                      0.065361   
sub-9930257                      0.056596   
sub-9931234                      0.050102   
sub-9939055                      0.085146   

                ctx_lh_caudalanteriorcingulate_thickness  \
participant_id                                             
sub-1002928                                     0.070757   
sub-1004359                                     0.077152   
sub-1016072                                     0.107828   
sub-1031654                                     0.076811   
sub-1072774                                     0.075648   
...                                                  ...   
sub-9889544                                     0.063802   
sub-9909448                                     0.116467   
sub-9930257                                     0.089051   
sub-9931234                                     0.107396   
sub-9939055                                     0.112179   

                ctx_lh_caudalmiddlefrontal_thickness  ctx_lh_cuneus_thickness  \
participant_id                                                                  
sub-1002928                                 0.045348                 0.131103   
sub-1004359                                 0.080876                 0.046602   
sub-1016072                                 0.115252                 0.146310   
sub-1031654                                 0.079381                 0.051086   
sub-1072774                                 0.045346                 0.077476   
...                                              ...                      ...   
sub-9889544                                 0.037739                 0.045298   
sub-9909448                                 0.081611                 0.086068   
sub-9930257                                 0.057623                 0.055001   
sub-9931234                                 0.058061                 0.087155   
sub-9939055                                 0.041061                 0.163699   

                ctx_lh_entorhinal_thickness  ctx_lh_fusiform_thickness  \
participant_id                                                           
sub-1002928                        0.090582                   0.031317   
sub-1004359                        0.079860                   0.047387   
sub-1016072                        0.103709                   0.076043   
sub-1031654                        0.083903                   0.122421   
sub-1072774                        0.082345                   0.032254   
...                                     ...                        ...   
sub-9889544                        0.073720                   0.063764   
sub-9909448                        0.089651                   0.043559   
sub-9930257                        0.106769                   0.101592   
sub-9931234                        0.090427                   0.055866   
sub-9939055                        0.092641                   0.058380   

                ctx_lh_inferiorparietal_thickness  \
participant_id                                      
sub-1002928                              0.051684   
sub-1004359                              0.043844   
sub-1016072                              0.052581   
sub-1031654                              0.063430   
sub-1072774                              0.027092   
...                                           ...   
sub-9889544                              0.037288   
sub-9909448                              0.036690   
sub-9930257                              0.058074   
sub-9931234                              0.046965

And that's it! You have everything you need to run the analysis on your own.

```{admonition} Advanced topic: Subsetting specific regions for averaging
:class: danger

In some cases, you may want to check the average imbalance in specific, pre-defined networks for instance. For instance, let's say you are investigating the **imbalance mapping** specifically in the temporal lobe. In such a case, you may be interested in computing the imbalance only in a specific set of regions. However, `sihnpy` doesn't yet natively support this. Two options are offered to you in this case:

1) Either you restrict your data (i.e., `thickness_data_final`) to include only the columns you are interested in.
2) You manipulate the `residual_array` to select only the parts of the rectangle that interest you.

The first solution is probably easiest, but it will require you to recompute the **imbalance mapping** for every different set of regions you want which may get computationally expensive when running it on many regions and/or combinations of regions.

The second solution is more complicated to do and require good knowledge of a [array slicing in Numpy](https://numpy.org/doc/stable/user/basics.indexing.html). It can take some getting used to, particularly when working in 3D. However, in this method, you only need to compute **imbalance mapping** once, and then you just need to feed the different slices to `imbalance_stats` which is much less computationally expensive.

If there is enough interest, I will adapt a way to do this natively in `sihnpy`. [Open an issue](https://github.com/stong3/sihnpy/issues) if you are interested!
```

### 4. Export

The next function needs a lot of arguments, but the goal is very simple: output the different measures computed during the **imbalance mapping** analysis. To output, simply follow the following command:

```python
imb.export(data=thickness_data_final, residual_array=residual_array, output_path="/path/to/output", avg_imb_by_region=avg_imb_by_region, avg_imb_by_person=avg_imb_by_person, avg_imb_by_pers_by_region=avg_imb_by_pers_by_region, name='test', all=False)
```

The only choice the user has over is whether we should export the matrices of orthogonal distances for each participant. This could be useful for other analyses but will require a lot more space. By default, this is not output, but the user can choose to do so by switching `False` to `True`.

## tl;dr

Too lazy to read everything? Or read everything and need a quick refresher? Here is the code in the order you need to make it work.

```python
from sihnpy.datasets import pad_imb_input #For practice data
from sihnpy import imbalance_mapping as imb #Imbalance mapping functions

volume_data, thickness_data, aseg_data = pad_imb_input() #Import the practice data
#If the data has more than 1 row by participant, clean the data
thickness_data_final = thickness_data[thickness_data['session'] == 'ses-BL00'].drop(labels=['session', 'run'], axis=1)

#Imbalance mapping
residual_array = imb.imbalance_mapping(data=thickness_data_final, type='abs')
avg_imb_by_region, avg_imb_by_person, avg_imb_by_pers_by_region = imb.imbalance_stats(data=thickness_data_final, residual_array=residual_array)

#Export
imb.export(data=thickness_data_final, residual_array=residual_array, output_path="/path/to/output", avg_imb_by_region=avg_imb_by_region, avg_imb_by_person=avg_imb_by_person, avg_imb_by_pers_by_region=avg_imb_by_pers_by_region, name='test', all=False)
```

## Appendix - Orthogonal Distance Regression (ODR)

**Imbalance mapping** in its original form[^Nadig_2021] leverages a special type of regression called **Orthogonal Distance Regression (ODR)** (also sometimes called total least square, or Denim regression). This is implemented by default in `sihnpy`'s function. However, `sihnpy` also exposes the function for users who would like to use it.

```{warning}
I should preface this section by saying that, while I have experience with principal component analysis (which is the basis of ODR in the form it is implemented), I am by no means a mathematician so some of the math involved go way over my head and the implementation involves some (or at lot) of trust in the original implementation. However, in my tests, results I got with `sihnpy` were entirely replicated using the original R library `pracma` as well as `scipy`'s ODR implementation which makes me more confident in its implementation here.
```

### Rationale

When doing a traditional regression--which refers most of the time to an Ordinary Least Square (OLS) regression--we try to **find the slope that minimizes the error between the outcome (Y) variable and the slope**. The goal here is to obtain the best possible slope describing the relationship between predictor and outcome. However, OLS assumes that there are only measurement errors on the **outcome variable**, which can become problematic when using an OLS regression between a predictor and outcome that can both have measurement error.[^Carr_2012]

This is problematic specifically in our case. When we do structural covariance, **we correlate the values in two brain regions, which may both have measurement error**. So how do we resolve this? 

Cue in **ODR**.

Instead of minimizing the distance between the outcome variable and the slope, **ODR minimizes the distance perpendicular to the slope on both the predictor and outcome**.[^Carr_2012],[^Pallavi_2022] This has the theoretical advantage of assuming that both the predictor and outcome have error[^Carr_2012] which is an attractive option in our case. But how does one figure out the slope **minimizing the orthogonal distance between the slope and the datapoints?** I'm glad you asked. A relatively easy way of doing this is to leverage **Principal Component Analysis (PCA)**.

In a PCA analysis, the goal is generally to **create new variables that can summarize variables in a dataset but conserve as much variance as possible**. This is so we can do what is called **dimensionality reduction**; we want to reduce the number of variables we need to use in our analyses but without losing important information (i.e. variance).[^Jolliffe_2016]

The PCA will always return the same number of components as the number of variables that are input to it. 




## References

Here are the references of this section:

[^Nadig_2021]:
[^Carr_2012]: James R. Carr (2012) Orthogonal regression: a teaching perspective, International Journal of Mathematical Education in Science and Technology, 43:1, 134-143, DOI: 10.1080/0020739X.2011.573876
[^Pallavi_2022]: Pallavi, Joshi, S., Singh, D. et al. Comprehensive Review of Orthogonal Regression and Its Applications in Different Domains. Arch Computat Methods Eng 29, 4027–4047 (2022). doi: 10.1007/s11831-022-09728-5
[^Joliffe_2016]: Jolliffe IT, Cadima J. Principal component analysis: a review and recent developments. Philos Trans A Math Phys Eng Sci. 2016 Apr 13;374(2065):20150202. doi: 10.1098/rsta.2015.0202. PMID: 26953178; PMCID: PMC4792409.